File: core/src/main/java/com/sequenceiq/cloudbreak/converter/CredentialToJsonConverter.java
package com.sequenceiq.cloudbreak.converter;

import static com.sequenceiq.cloudbreak.cloud.model.Platform.platform;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import org.springframework.stereotype.Component;

import com.sequenceiq.cloudbreak.api.model.CredentialResponse;
import com.sequenceiq.cloudbreak.domain.Credential;
import com.sequenceiq.cloudbreak.service.stack.resource.definition.credential.CredentialDefinitionService;
import com.sequenceiq.cloudbreak.service.topology.TopologyService;

@Component
public class CredentialToJsonConverter extends AbstractConversionServiceAwareConverter<Credential, CredentialResponse> {
    private static final List<String> FIELDS_TO_COVER = Arrays.asList("password", "secretKey", "serviceAccountPrivateKey");

    private static final String PLACEHOLDER = "********";

    @Inject
    private CredentialDefinitionService credentialDefinitionService;

    @Inject
    private TopologyService topologyService;

    @Override
    public CredentialResponse convert(Credential source) {
        CredentialResponse credentialJson = new CredentialResponse();
        credentialJson.setId(source.getId());
        credentialJson.setCloudPlatform(source.cloudPlatform());
        credentialJson.setName(source.getName());
        credentialJson.setPublicInAccount(source.isPublicInAccount());
        if (source.getAttributes() != null) {
            Map<String, Object> parameters =
                    credentialDefinitionService.revertAndRemoveProperties(platform(source.cloudPlatform()), source.getAttributes().getMap());
            credentialJson.setParameters(parameters);
        }
        credentialJson.setDescription(source.getDescription() == null ? "" : source.getDescription());
        credentialJson.setPublicKey(source.getPublicKey());
        credentialJson.setLoginUserName(source.getLoginUserName());
        if (source.getTopology() != null) {
            credentialJson.setTopologyId(source.getTopology().getId());
        }
        coverSensitiveData(credentialJson);
        return credentialJson;
    }

    private void coverSensitiveData(CredentialResponse response) {
        for (String field : FIELDS_TO_COVER) {
            if (response.getParameters().get(field) != null) {
                response.getParameters().put(field, PLACEHOLDER);
            }
        }
    }
}

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/CredentialToJsonConverter.kt
package com.sequenceiq.cloudbreak.converter

import com.sequenceiq.cloudbreak.cloud.model.Platform.platform

import javax.inject.Inject

import org.springframework.stereotype.Component

import com.sequenceiq.cloudbreak.api.model.CredentialResponse
import com.sequenceiq.cloudbreak.domain.Credential
import com.sequenceiq.cloudbreak.service.stack.resource.definition.credential.CredentialDefinitionService
import com.sequenceiq.cloudbreak.service.topology.TopologyService

@Component
class CredentialToJsonConverter : AbstractConversionServiceAwareConverter<Credential, CredentialResponse>() {

    @Inject
    private val credentialDefinitionService: CredentialDefinitionService? = null
    @Inject
    private val topologyService: TopologyService? = null

    override fun convert(source: Credential): CredentialResponse {
        val credentialJson = CredentialResponse()
        credentialJson.id = source.id
        credentialJson.cloudPlatform = source.cloudPlatform()
        credentialJson.name = source.name
        credentialJson.isPublicInAccount = source.isPublicInAccount
        if (source.attributes != null) {
            val parameters = credentialDefinitionService!!.revertProperties(Companion.platform(source.cloudPlatform()), source.attributes.map)
            credentialJson.parameters = parameters
        }
        credentialJson.description = if (source.description == null) "" else source.description
        credentialJson.publicKey = source.publicKey
        credentialJson.loginUserName = source.loginUserName
        if (source.topology != null) {
            credentialJson.topologyId = source.topology.id
        }
        clearPasswordField(credentialJson)
        return credentialJson
    }

    private fun clearPasswordField(response: CredentialResponse) {
        if (response.parameters[PASSWORD_FIELD] != null) {
            response.parameters.put(PASSWORD_FIELD, PASSWORD_PLACEHOLDER)
        }
    }

    companion object {
        private val PASSWORD_FIELD = "password"
        private val PASSWORD_PLACEHOLDER = "********"
    }
}
